File: 1676-lowest-common-ancestor-of-a-binary-tree-iv/1676-lowest-common-ancestor-of-a-binary-tree-iv.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', nodes: 'List[TreeNode]') -> 'TreeNode':

        def dfs(node):
            if not node or node in nodes:
                return node
            left = dfs(node.left)
            right = dfs(node.right)
            if not left:
                return right
            if not right:
                return left
            return node
        return dfs(root)
        

File: 1676-lowest-common-ancestor-of-a-binary-tree-iv/1676-lowest-common-ancestor-of-a-binary-tree-iv.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    HashSet<Integer> table = new HashSet<>();
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode[] nodes) {
        for (TreeNode node : nodes){
            table.add(node.val);
        }
        return find(root);
    }
    public TreeNode find(TreeNode node){
        if (node == null || table.contains(node.val)){
            return node;
        }
        TreeNode left = find(node.left);
        TreeNode right = find(node.right);
        if (left == null){
            return right ;
        }
        if (right == null){
            return left;
        }
        return node;
    }
}